packages/dds/tree/docs/dataModel/README.md

```diff
@@ -0,0 +1,128 @@
+# Data Model
+
+## Introduction
+
+This document discusses how tree data is conceptually organized within the SharedTree DDS. It is primarily of interested to Fluid developers working on SharedTree.  Most SharedTree users will interact with tree data through a pre-existing API layer.  However, the data model may be of interest to advanced users who are implementing a specialized API (see #8989).
+
+The SharedTree data model is closely related to, but not precisely the same as:
+
+- The internal in-memory representation of tree data
+- The serialized tree data sent across the network
+- The format used to persist tree data in storage
+
+While each of the above involve specific encodings of the SharedTree data model, the SharedTree data model itself is not concerned with the byte-level representation.
+
+## Requirements
+
+### JSON Interoperability
+
+We take it as a given that the underlying data model for the SharedTree is tree-structured data. We also agree that JSON is the modern lingua franca of the web and services. We therefore begin with the following requirements:
+
+1. JSON must efficiently and losslessly round-trip to and from the underlying SharedTree data model, with some caveats.
```

Comment:
"JSON must [...] losslessly round-trip to and from the underlying SharedTree data model" is ambiguous.

Ignoring the performance aspects and just focusing on the correctness ones:

I think it means one of the following things, but it is unclear which:
1. There exists an injective function from our trees to json and an injective function from our json to our trees.
2. There exists a bijection  between our trees and json (aka the above, but also requiring these functions be each others inverses)

I think we want something closer to the first one, but more specifically I think we want:
1. Fluid tree must be able to be losslessly exported to JSON, and should aim to make the produced JSON as friendly to read as possible. (Aka: there is an injection from our tree to JSON, where an outside observer would easily interpret the JSON with semantics that align with our tree's semantics)
2. JSON data can losslessly be can be encoded in our trees (excepting non float64 representable numbers) such that our viewing and editing APIs seem nature to someone familiar with the original JSON's format.
3. When working with strongly typed data (with a schema), there should an intuitive bijection between the tree data using those types, and a well defined (and expressible with JsonSchema other than the exclusion of non float64 representable numbers) subset of JSON.
4. Deviations from JSON in the underlying SharedTree data model must be well justified.

Reply:
Unfortuntaley, you cannot have a (perfect) bijection and at the same time extend the data model (add types, augmentations and identities).

I guess the ideal would be, to have any deviations from JSON be optional from a users perspective. If the user doesn't care about those (and doesn't opt-in / explictly uses the JSON API) the system works like JSON and the mapping is bijective. The user won't see extensions of the data model, but they should be preserved as much as possible (however, if the user downloads the JSON representation and then reuploads it, there would be no way to preserve the attached information).  Where possible, extensions should also be automatically inserted into the data (types derived from the schema and IDs assigned for new nodes).

There are two different cases:
1. The user uses fluid via the SDK and its API. In this case, extensions of the data model could be additional APIs, which don't cause problems, but can be ignored by the user (e.g. a `getID` or `getType` function or fields that can be accessed via symbols).
2. The user uses an interface that actually returns/receives a JSON. This potentially would apply to exporting/serializing from the internal representation via a function, but the more relevant use case would be a REST server endpoint. I think in such a case, the user would have to specifiy, whether he wants to see pure JSON (which then provides a bijection) or wants to see some of the extensions of the data model (via a parameter to the serialization function or a query string or HTTP header in the server interface). If extensions are enabled, this will make it necessary to add new fields into the JSON and as a consequence also to use some escaping mechanism to avoid name collisions, still it should as much as possible preserve the original structure of the JSON (the most difficult case here are probably augmentation on primitive value leaf nodes).

A more complex case here are APIs that modify the data via an (unextended) JSON. E.g. a PUT into an existing document. Here, we would have the problem that potential extensions would be lost, if we just overwrite from the JSON that does not have those (plus the problem that overwriting causes problems with collaboration). Essentially there are two options:
* Just overwriting, accepting that this is a destructive operation
* Converting to an edit operation by diffing and then applying only the diff. This would preserve most extensions and would potentially work better with collaboration, but could have unexpected effects (since the user won't be able to explicitly declare how to merge changes).

I think we should support this use-case, due to the demand for simple APIs by users (again we could let the user choose via an HTTP header/query parameter whether the diff approach or a complete overwrite should be used), but should make them aware that the better pattern is, to use a PATCH call (or the Fluid SDK), since those better preserve the intention of an edit.

Reply:
@CraigMacomber, @ruiterr - A added a new 'Potential JSON Caveats' section that addresses two points of feedback:

1.  It calls out the caveat around numbers.
2.  It explains that the mapping from JSON -> SharedTree is injective (not bijective), and therefore going from SharedTree -> JSON will require tradeoffs.  One possible tradeoff is the use of reserved key spaces, which impacts JSON interoperability.